### PR TITLE
Allow multiple mutators for a single admission controller

### PR DESCRIPTION
**Pull request by croomes**
This allows an admission controller to run multiple mutators for different tasks.

For Pod creation, this could be to add the StorageOS scheduler and to set a failure mode or failure mode label that should be applied to all of the Pod's PVCs.

This pattern will be used for PVC create where we will have separate mutators for encryption and labels.

### CODE REVIEW

File: controllers/pod-mutator/scheduler/scheduler.go [hunk 1]
```diff
@@ -101,9 +65,9 @@ func (p *PodSchedulerSetter) mutatePodsFn(ctx context.Context, pod *corev1.Pod,
 
 	// Skip mutation if the pod annotation has false schedule annotation.
 	if val, exists := pod.ObjectMeta.Annotations[p.SchedulerAnnotationKey]; exists {
-		boolVal, err := strconv.ParseBool(val)
+		enabled, err := strconv.ParseBool(val)
 		// No error in parsing and the value is false, skip the pod.
-		if err == nil && !boolVal {
+		if err == nil && !enabled {
 			log.V(4).Info("pod has storageos.com/scheduler=false annotaion, skipping")
```
Comment (avestuk) [inline controllers/pod-mutator/scheduler/scheduler.go:71]:
```suggestion
			log.V(4).Infof("pod has %s=false annotaion, skipping", p.SchedulerAnnotationKey)
```

Does it make sense to use the const you have defined already? I suppose there's not much chance of it changing. Really I'm just curious about whether you chose to not use the const.

File: controllers/pod-mutator/controller.go [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+	if err != nil {
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+
+	// Get the request namespace. This is needed because sometimes the decoded
+	// pod object lacks namespace info.
+	namespace := req.AdmissionRequest.Namespace
+
+	// Create a copy of the pod to mutate.
+	copy := pod.DeepCopy()
```
Comment (avestuk) [inline controllers/pod-mutator/controller.go:56]: Why do you have to create a copy of the pod as opposed to just modifying the `pod` object directly and sending that back?

**Review by avestuk — APPROVED** [review]

File: controllers/pod-mutator/README.md [hunk 3]
```diff
@@ -1,41 +1,35 @@
...
+instances of the api-manager share the same certificate and will check for
 updates periodically, configured with the `-webhook-cert-refresh-interval` flag.
 
 Certificates are valid for 1 year and re-issued after 6 months.  The
 `-webhook-cert-refresh-interval` should be kept to run frequently (default
 `30m`) as restarting the api-manager will reset the refresh timer.
 
+## Disabling
+
+It is not possible to disable the Webhook server.  Instead, disable thegla
```
Comment (mhmxs) [inline controllers/pod-mutator/README.md:30]: `thegla` looks a typo here.

File: controllers/pod-mutator/scheduler/README.md [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+not have the `storageos.com/scheduler=false` annotation set on the Pod.
+
+## Trigger
+
+All Pod create events are evaluated.
+
+Pods that have the annotation `storageos.com/scheduler=false` set are ignored.
+
+The Pod's volumes are enumerated, and if at least one volume is backed by
+StorageOS, the SchedulerName will be changed to use the StorageOS Pod scheduler.
```
Comment (mhmxs) [inline controllers/pod-mutator/scheduler/README.md:14]: SchedulerName should be `highlighted`

File: controllers/pod-mutator/controller.go (hunk 2 above)
Comment (darkowlzz) [reply controllers/pod-mutator/controller.go:56]: This code is based on https://github.com/storageos/cluster-operator/blob/v2.3.4/internal/pkg/admission/scheduler/handler.go#L64 where we used to pass the pod object to `PatchResponse` at the very end of the function to return a response. In the latest webhook package of controller runtime, `PatchResponse` is replaced with `PatchResponseFromRaw` which accepts the raw object in the request. So, we avoid doing a deep copy here.

File: controllers/pod-mutator/scheduler/README.md [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+## Tunables
+
+Default values work well when the api-manager is installed by the
+cluster-operator.  They should only be changed under advisement from StorageOS
+support.
+
+`-scheduler-name` is the name of the scheduler extender to set in the Pod's
+`SchedulerName` field.  If set to an empty string, no updates will be made. When
+set, the scheduler name must match the name of the scheduler extender configured
+in the StorageOS cluster-operator. (default "storageos-scheduler)".
```
Comment (darkowlzz) [inline controllers/pod-mutator/scheduler/README.md:33]:
```suggestion
in the StorageOS cluster-operator. (default "storageos-scheduler").
```

File: controllers/pod-mutator/scheduler/scheduler.go [hunk 6]
```diff
@@ -37,59 +34,26 @@ type PodSchedulerSetter struct {
...
-	return admission.PatchResponseFromRaw(req.Object.Raw, marshaledPod)
-}
-
-// mutatePodFn mutates a given pod with a configured scheduler name if the pod
-// is associated with volumes managed by the configured provisioners.
-func (p *PodSchedulerSetter) mutatePodsFn(ctx context.Context, pod *corev1.Pod, namespace string) error {
+// MutatePod mutates a given pod with a configured scheduler name if the pod is
+// associated with volumes managed by the configured provisioners.
+//
+// Errors returned here will block creation of the Pod.
```
Comment (darkowlzz) [inline controllers/pod-mutator/scheduler/scheduler.go:55]: May not be accurate. It depends on the webhook configuration if the failure is ignored or respected.

**Comment by darkowlzz** [comment]
The implementation looks good to me. Left some inline comments about minor improvements.

File: controllers/pod-mutator/scheduler/scheduler.go (hunk 1 above)
Comment (croomes) [reply controllers/pod-mutator/scheduler/scheduler.go:71]: yeah, it probably does.  With `logr` they recommend against vars in the message and encourage k=v pairs instead, but that doesn't work well here.  Changed.

File: controllers/pod-mutator/controller.go (hunk 2 above)
Comment (croomes) [reply controllers/pod-mutator/controller.go:56]: Removed the `DeepCopy()`

File: controllers/pod-mutator/scheduler/scheduler.go (hunk 6 above)
Comment (croomes) [reply controllers/pod-mutator/scheduler/scheduler.go]: :+1: Changed wording

File: controllers/test_helpers.go [hunk 7]
```diff
@@ -0,0 +1,45 @@
+package controllers
```
Comment (mhmxs) [inline controllers/test_helpers.go:1]: In this way test related codebase looks like production code. I suggest to refactor the code to ensure test functions are only available in tests. If it is not possible a common solution to create a `controllertest` package for separation.

File: controllers/test_helpers.go [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+	if m.error {
+		return errors.New("error")
+	}
+	if obj.Annotations == nil {
+		obj.Annotations = make(map[string]string)
+	}
+	obj.Annotations[m.key] = m.value
+	return nil
+}
+func (m testMutator) MutatePVC(ctx context.Context, obj *corev1.PersistentVolumeClaim, namespace string) error {
```
Comment (mhmxs) [inline controllers/test_helpers.go:36]: Missing new line :)

**Review by mhmxs — APPROVED** [review]

File: controllers/test_helpers.go (hunk 7 above)
Comment (mhmxs) [reply controllers/test_helpers.go:1]: Sorry, my bad, this package contains only tests :D

**Review by darkowlzz — APPROVED** [review]
Other than the remaining minor comments and squashing the commits, this LGTM!
Thanks.

File: controllers/test_helpers.go (hunk 8 above)
Comment (croomes) [reply controllers/test_helpers.go:37]: done, thanks

**Pull request merged**